File: src/main/java/net/wyvest/wyvtilities/commands/WyvtilsCommands.java
package net.wyvest.wyvtilities.commands;

import net.minecraft.util.EnumChatFormatting;
import net.wyvest.wyvtilities.Wyvtilities;
import net.wyvest.wyvtilities.config.WyvtilsConfig;
import net.wyvest.wyvtilities.utils.GexpUtils;
import xyz.matthewtgm.json.util.JsonApiHelper;
import xyz.matthewtgm.tgmlib.commands.advanced.Command;
import xyz.matthewtgm.tgmlib.util.GuiHelper;
import xyz.matthewtgm.tgmlib.util.Multithreading;
import xyz.matthewtgm.tgmlib.util.Notifications;

@SuppressWarnings("unused")
@Command(name = "wyvtilities", aliases = {"wyvtils", "wytils"}, tabCompleteOptions = {"help", "config", "setkey", "gexp"})
public class WyvtilsCommands {
    @Command.Process
    public void process(String[] args) {
        if (args.length <= 0) {
            GuiHelper.open(WyvtilsConfig.INSTANCE.gui());
        } else {
            Wyvtilities.INSTANCE.sendMessage(EnumChatFormatting.RED + "Unknown argument. Type /wyvtilities help for correct usage.");
        }
    }

    @Command.Argument(name = "help")
    public void help() {
        Wyvtilities.INSTANCE.sendHelpMessage();
    }

    @Command.Argument(name = "config")
    public void config() {
        GuiHelper.open(WyvtilsConfig.INSTANCE.gui());
    }

    @Command.Argument(name = "setkey")
    public void setKey(String[] args) {
        try {
            if (args.length == 1 && JsonApiHelper.getJsonObject("https://api.hypixel.net/key?key=" + args[1])
                    .get("success").getAsBoolean()
            ) {
                WyvtilsConfig.apiKey = args[1];
                WyvtilsConfig.INSTANCE.markDirty();
                WyvtilsConfig.INSTANCE.writeData();
                Wyvtilities.INSTANCE.sendMessage(EnumChatFormatting.GREEN.toString() + "Saved API key successfully!");
            } else {
                Wyvtilities.INSTANCE.sendMessage(EnumChatFormatting.RED + "Invalid API key! Please try again.");
            }
        } catch (Exception ex) {
            Wyvtilities.INSTANCE.sendMessage(EnumChatFormatting.RED + "Invalid API key! Please try again.");
            ex.printStackTrace();
        }
    }

    @Command.Argument(name = "gexp")
    public void getGexp(String[] args) {
        if (WyvtilsConfig.apiKey.isEmpty()) {
            Wyvtilities.INSTANCE.sendMessage(EnumChatFormatting.RED + "You need to provide a valid API key to run this command! Type /api new to autoset a key.");
        } else {
            if (args.length <= 1) {
                Multithreading.runAsync(() -> {
                    GexpUtils.INSTANCE.getGEXP();
                    Notifications.push("Wyvtilities", "You currently have " + GexpUtils.gexp + " guild EXP.");
                });
            } else {
                Multithreading.runAsync(() -> {
                    GexpUtils.INSTANCE.getGEXP(args[1]);
                    Notifications.push("Wyvtilities", args[1] + " currently have " + GexpUtils.gexp + " guild EXP.");
                });
            }
        }
    }

}

File: src/main/kotlin/net/wyvest/wyvtilities/commands/WyvtilsCommands.kt
/*
 * Wyvtilities - Utilities for Hypixel 1.8.9.
 * Copyright (C) 2021 Wyvtilities
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package net.wyvest.wyvtilities.commands

import gg.essential.api.EssentialAPI
import gg.essential.api.commands.Command
import gg.essential.api.commands.DefaultHandler
import gg.essential.api.commands.DisplayName
import gg.essential.api.commands.SubCommand
import gg.essential.api.utils.Multithreading
import net.wyvest.wyvtilities.config.WyvtilsConfig
import net.wyvest.wyvtilities.utils.HypixelUtils

@Suppress("unused")
object WyvtilsCommands : Command("wyvtilities", true) {

    override val commandAliases = setOf(Alias("wyvtils"), Alias("wytils"), Alias("wyvtil"), Alias("wytil"), Alias("wyvest")) //dont ask about the wyvest alias

    @DefaultHandler
    fun handle() {
        EssentialAPI.getGuiUtil().openScreen(WyvtilsConfig.gui())
    }

    @SubCommand("config", description = "Opens the config GUI for Wyvtils")
    fun config() {
        EssentialAPI.getGuiUtil().openScreen(WyvtilsConfig.gui())
    }

    @SubCommand("gexp", description = "Gets the GEXP of the player specified")
    fun gexp(@DisplayName("username") username: String?, @DisplayName("type") type: String?) {
        try {
            Multithreading.runAsync {
                if (username != null) {
                    if (type == null) {
                        if (HypixelUtils.getGEXP(username)) {
                            EssentialAPI.getNotifications()
                                .push("Wyvtilities", "$username currently has " + HypixelUtils.gexp + " guild EXP.")
                        } else {
                            EssentialAPI.getNotifications()
                                .push("Wyvtilities", "There was a problem trying to get $username's GEXP.")
                        }
                    } else {
                        if (type == "daily") {
                            if (HypixelUtils.getGEXP(username)) {
                                EssentialAPI.getNotifications()
                                    .push("Wyvtilities", "$username currently has " + HypixelUtils.gexp + " daily guild EXP.")
                            } else {
                                EssentialAPI.getNotifications()
                                    .push("Wyvtilities", "There was a problem trying to get $username's daily GEXP.")
                            }
                        } else if (type == "weekly") {
                            if (HypixelUtils.getWeeklyGEXP(username)) {
                                EssentialAPI.getNotifications()
                                    .push("Wyvtilities", "$username currently has " + HypixelUtils.gexp + " weekly guild EXP.")
                            } else {
                                EssentialAPI.getNotifications()
                                    .push("Wyvtilities", "There was a problem trying to get $username's weekly GEXP.")
                            }
                        } else {
                            EssentialAPI.getNotifications()
                                .push("Wyvtilities", "The type argument was not valid.")
                        }
                    }
                } else {
                    if (HypixelUtils.getGEXP()) {
                        EssentialAPI.getNotifications()
                            .push("Wyvtilities", "You currently have " + HypixelUtils.gexp + " guild EXP.")
                    } else {
                        EssentialAPI.getNotifications()
                            .push("Wyvtilities", "There was a problem trying to get your GEXP.")
                    }
                }
            }
        } catch (e: NullPointerException) {
            e.printStackTrace()
        }
    }

    @SubCommand("winstreak", description = "Gets the winstreak of the player specified")
    fun winstreak(@DisplayName("username") username: String?, @DisplayName("gamemode") gamemode: String?) {
        try {
            Multithreading.runAsync {
                if (username != null) {
                    if (gamemode != null) {
                        if (HypixelUtils.getWinstreak(username, gamemode)) {
                            EssentialAPI.getNotifications()
                                .push(
                                    "Wyvtilities",
                                    "$username currently has a " + HypixelUtils.winstreak + " winstreak in $gamemode."
                                )
                        } else {
                            EssentialAPI.getNotifications()
                                .push(
                                    "Wyvtilities",
                                    "There was a problem trying to get $username's winstreak in $gamemode."
                                )
                        }
                    } else {
                        if (HypixelUtils.getWinstreak(username)) {
                            EssentialAPI.getNotifications()
                                .push(
                                    "Wyvtilities",
                                    "$username currently has a " + HypixelUtils.winstreak + " winstreak."
                                )
                        } else {
                            EssentialAPI.getNotifications()
                                .push("Wyvtilities", "There was a problem trying to get $username's winstreak.")
                        }
                    }
                } else {
                    if (HypixelUtils.getWinstreak()) {
                        EssentialAPI.getNotifications()
                            .push("Wyvtilities", "You currently have a " + HypixelUtils.winstreak + " winstreak.")
                    } else {
                        EssentialAPI.getNotifications()
                            .push("Wyvtilities", "There was a problem trying to get your winstreak.")
                    }
                }
            }
        } catch (e: NullPointerException) {
            e.printStackTrace()
        }
    }

}